frontend/apps/reader/modules/readerfooter.lua

```diff
@@ -2446,16 +2446,9 @@ end
 ReaderFooter.onNetworkDisconnected = ReaderFooter.onNetworkConnected
 
 function ReaderFooter:onSwapPageTurnButtons()
-    if self.settings.page_turning_inverted then
-        self:maybeUpdateFooter()
-    end
-end
-
-function ReaderFooter:onToggleReadingOrder()
-    if self.settings.page_turning_inverted then
-        self:maybeUpdateFooter()
-    end
+    UIManager:scheduleIn(0.1, self.autoRefreshFooter)
```

Comment:
Nope.

Reply:
Nope nope nope nope.

Reply:
one was enough but, what's the matter?

Reply:
This isn't how you do this, which tells me you're trying to fit a square peg into a round hole ;).

So: what is this trying to handle (i.e., where does that event emanate, and what's on screen at the time).

Reply:
> So: what is this trying to handle

when you use a gesture to trigger the page-turn button swap or the zone swap thing, update the status bar.

> and what's on screen at the time)?

the book one is reading.

Reply:
Oh, ReaderFooter receives the event before DeviceListener, so the settings aren't actually swapped yet -_-".

Reply:
> Oh, ReaderFooter receives the event before DeviceListener, so the settings aren't actually swapped yet -_-".

so that is why the maybeUpdateFooter() wouldn't work. Nice. I was losing my mind about it.

you also said, we should add it to callback for the menu ones, `UIManager:nextTick(xxx)` what would that xxx look like? 

https://github.com/koreader/koreader/blob/7e52c15de29c013413f3d9ffef37d2de86b44dbf/frontend/ui/elements/physical_buttons.lua#L48-L51

Reply:
If the Menu is not actually in ReaderFooter, I'd rather said menu switch to events instead of reimplementing the same thing twice ;).

Reply:
wait, are you saying: make those two individual page turn inversions an event on DeviceListener like onSwapPageTurnButtons?

Reply:
No, I'm saying make the menu use the events instead of duplicating what the devicelistener handler does (I assume).

Reply:
mmm, but that also doesn't update the status bar...

Reply:
Something along the lines of

```diff
diff --git a/frontend/apps/reader/modules/readerfooter.lua b/frontend/apps/reader/modules/readerfooter.lua
index 255400087..7ec8d8ee4 100644
--- a/frontend/apps/reader/modules/readerfooter.lua
+++ b/frontend/apps/reader/modules/readerfooter.lua
@@ -2445,6 +2445,14 @@ function ReaderFooter:onNetworkConnected()
 end
 ReaderFooter.onNetworkDisconnected = ReaderFooter.onNetworkConnected
 
+function ReaderFooter:onSwapPageTurnButtons()
+    if self.settings.page_turning_inverted then
+        -- We may receive the event *before* DeviceListener, so delay this to make sure it had a chance to actually swap the settings.
+        UIManager:nextTick(self.maybeUpdateFooter, self)
+    end
+end
+ReaderFooter.onToggleReadingOrder = ReaderFooter.onSwapPageTurnButtons
+
 function ReaderFooter:onSetRotationMode()
     self:updateFooterContainer()
     self:resetLayout(true)
diff --git a/frontend/device/devicelistener.lua b/frontend/device/devicelistener.lua
index 8481f7670..456b47e94 100644
--- a/frontend/device/devicelistener.lua
+++ b/frontend/device/devicelistener.lua
@@ -310,15 +310,25 @@ function DeviceListener:onToggleFlashOnPagesWithImages()
     G_reader_settings:flipNilOrTrue("refresh_on_pages_with_images")
 end
 
-function DeviceListener:onSwapPageTurnButtons(show_notification)
+function DeviceListener:onSwapPageTurnButtons(show_notification, side)
     G_reader_settings:flipNilOrFalse("input_invert_page_turn_keys")
-    Device:invertButtons()
+    local key_name
+    if side == "left" then
+        Device:invertButtonsLeft()
+        key_name = _("Left-side page-turn buttons")
+    elseif side == "right" then
+        Device:invertButtonsRight()
+        key_name = _("Right-side page-turn buttons")
+    else
+        Device:invertButtons()
+        key_name = _("Page-turn buttons")
+    end
     if show_notification then
         local new_text
         if G_reader_settings:isTrue("input_invert_page_turn_keys") then
-            new_text = _("Page-turn buttons inverted.")
+            new_text = key_name .. " " .. _("inverted.")
         else
-            new_text = _("Page-turn buttons no longer inverted.")
+            new_text = key_name .. " " .. _("no longer inverted.")
         end
         Notification:notify(new_text)
     end
diff --git a/frontend/dispatcher.lua b/frontend/dispatcher.lua
index ea71d2f1d..ccba3c63e 100644
--- a/frontend/dispatcher.lua
+++ b/frontend/dispatcher.lua
@@ -80,7 +80,9 @@ local settingsList = {
     touch_input_off = {category="none", event="IgnoreTouchInput", arg=true, title=_("Disable touch input"), device=true},
     toggle_touch_input = {category="none", event="IgnoreTouchInput", title=_("Toggle touch input"), device=true, separator=true},
     ----
-    swap_page_turn_buttons = {category="none", event="SwapPageTurnButtons", arg=true, title=_("Invert page turn buttons"), device=true, condition=Device:hasKeys(), separator=true},
+    swap_left_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args={true, "left"}, title=_("Invert left-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
+    swap_right_page_turn_buttons = {category="none", event="SwapPageTurnButtons", args={true, "right"}, title=_("Invert right-side page-turn buttons"), device=true, condition= Device:hasDPad() and Device:useDPadAsActionKeys()},
+    swap_page_turn_buttons = {category="none", event="SwapPageTurnButtons", arg=true, title=_("Invert page-turn buttons"), device=true, condition=Device:hasKeys(), separator=true},
     ----
     toggle_key_repeat = {category="none", event="ToggleKeyRepeat", title=_("Toggle key repeat"), device=true, condition=Device:hasKeys() and Device:canKeyRepeat(), separator=true},
     toggle_gsensor = {category="none", event="ToggleGSensor", title=_("Toggle accelerometer"), device=true, condition=Device:hasGSensor()},
@@ -306,6 +308,8 @@ local dispatcher_menu_order = {
     "touch_input_off",
     "toggle_touch_input",
     ----
+    "swap_left_page_turn_buttons",
+    "swap_right_page_turn_buttons",
     "swap_page_turn_buttons",
     ----
     "toggle_key_repeat",
diff --git a/frontend/ui/elements/physical_buttons.lua b/frontend/ui/elements/physical_buttons.lua
index 87c2a1e49..4f364d037 100644
--- a/frontend/ui/elements/physical_buttons.lua
+++ b/frontend/ui/elements/physical_buttons.lua
@@ -25,7 +25,7 @@ local PhysicalButtons = {
 
 if Device:hasDPad() and Device:useDPadAsActionKeys() then
     table.insert(PhysicalButtons.sub_item_table, {
-        text = _("Invert left-side page turn buttons"),
+        text = _("Invert left-side page-turn buttons"),
         enabled_func = function()
             return not G_reader_settings:isTrue("input_invert_page_turn_keys")
         end,
@@ -34,11 +34,11 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
         end,
         callback = function()
             G_reader_settings:flipNilOrFalse("input_invert_left_page_turn_keys")
-            Device:invertButtonsLeft()
+            UIManager:broadcastEvent(Event:new("SwapPageTurnButtons", false, "left"))
         end,
     })
     table.insert(PhysicalButtons.sub_item_table, {
-        text = _("Invert right-side page turn buttons"),
+        text = _("Invert right-side page-turn buttons"),
         enabled_func = function()
             return not G_reader_settings:isTrue("input_invert_page_turn_keys")
         end,
@@ -47,7 +47,7 @@ if Device:hasDPad() and Device:useDPadAsActionKeys() then
         end,
         callback = function()
             G_reader_settings:flipNilOrFalse("input_invert_right_page_turn_keys")
-            Device:invertButtonsRight()
+            UIManager:broadcastEvent(Event:new("SwapPageTurnButtons", false, "right"))
         end,
         separator = true,
     })
```

Reply:
Tweaked the delay above as is was screwing with Menu highlights.

Reply:
`UIManager:scheduleIn(0.5, self.maybeUpdateFooter, self)`

nope nope nope nope. (sorry I couldn't resist... ;) )

I'll try it in a bit, will get back soon. the only thing I see from looking at it is, swapping one side [only] would now flip the G-setting for swapping both sides, is that intentional?

Reply:
Oh, I may have forgotten to handle the settings properly, yeah ;D

Reply:
Yup, definitely ;p.